plugins/osdn/vnids.go

```diff
@@ -88,44 +185,42 @@ func (oc *OsdnController) assignVNID(namespaceName string) error {
 		}
 		return err
 	}
-	oc.VNIDMap[namespaceName] = netid
-	log.Infof("Assigned id %d to namespace %q", netid, namespaceName)
+	oc.setVNID(namespaceName, netid)
 	return nil
 }
 
 func (oc *OsdnController) revokeVNID(namespaceName string) error {
-	err := oc.Registry.DeleteNetNamespace(namespaceName)
+	// Remove NetID from vnid map
+	netid_found := true
+	netid, err := oc.unSetVNID(namespaceName)
 	if err != nil {
-		return err
+		log.Error(err)
+		netid_found = false
 	}
-	netid, found := oc.VNIDMap[namespaceName]
-	if !found {
-		return fmt.Errorf("Error while fetching Net ID for namespace: %s", namespaceName)
+
+	// Delete NetNamespace object
+	err = oc.Registry.DeleteNetNamespace(namespaceName)
+	if err != nil {
+		return err
 	}
-	delete(oc.VNIDMap, namespaceName)
 
-	// Skip AdminVNID as it is not part of Net ID allocation
-	if netid == AdminVNID {
+	// Skip NetID release if
+	// - Value matches AdminVNID as it is not part of NetID allocation or
+	// - NetID is not found in the vnid map
+	if (netid == AdminVNID) || !netid_found {
 		return nil
 	}
 
 	// Check if this netid is used by any other namespaces
 	// If not, then release the netid
-	netid_inuse := false
-	for name, id := range oc.VNIDMap {
-		if id == netid {
-			netid_inuse = true
-			log.V(5).Infof("Net ID %d for namespace %q is still in use by namespace %q", netid, namespaceName, name)
-			break
-		}
-	}
-	if !netid_inuse {
+	if !oc.checkVNID(netid) {
 		err = oc.netIDManager.ReleaseNetID(netid)
```

Comment:
Are API calls synchronized into a single thread or is it possible for two to be happening at the same time in different goroutines? If the latter, then it's possible that checkVNID() will return false, but then another namespace will be modified to use netid in another thread before the call to ReleaseNetID(). You need to hold the mutex across both calls in order to be safe.